### ISSUE TITLE
"Failed to acquire global lock" after several syncs

### ISSUE THREAD

**Issue by darkpixel** [bolthole/zrep#8]
After syncing a few times, zrep starts reporting that it "Failed to acquire global lock".

```
root@usvansdnas01:~# zrep -t zrep-local sync all
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
ln: failed to create symbolic link ‘/var/run/zrep.lock’: File exists
Failed to acquire global lock
Error: Cannot lock tank/officeshare. Cannot continue

root@usvansdnas01:~# ls -lha /var/run/zrep.lock 
lrwxrwxrwx 1 root root 10 Sep 10 18:35 /var/run/zrep.lock -> /proc/6479
root@usvansdnas01:~# ps uax | grep 6479
root     14552  0.0  0.0  12720  2112 pts/2    S+   11:18   0:00 grep 6479
root@usvansdnas01:~# 
```

Is it possible to have zrep override the lock when the process referenced no longer exists?

**Comment by ppbrown** [comment]
I have two fairly strong reasons why I'm reluctant to do that.

1. it is a good indication that something is very, VERY wrong with your
system.
This almost never happens. So when it does, you should probably look into
it more.


2. This script runs across now an almost uncountable number of slightly
different OS's :-)
I dont know of a 100% reliable way to check for "does this process still
exist?" across all the OS's.

**Comment by darkpixel** [comment]
1. As far as I can tell, nothing is wrong.  ;)  I'm not sure why the lock file is getting left behind.  My initial guess would be bad network connectivity causing the replication to time out.  I guess I need to turn on -x again and dig through the logs after it fails.

2. You already appear to be creating a symlink to /proc/$pid.  Can't you do something like ```if [ -e /var/run/zrep.lock ]``` (excuse my poor bash skills) to see if the target of the symlink still exists?  On linux the /proc/$pid folder is removed when the process goes away.

**Comment by ppbrown** [comment]
Hmm.
my caution was going from memory, where I think someone complained to me
that their system did not have /proc


however.. I see in the code that it DOES still attempt to remove stale
locks.
(just like I thought it did :)

in zrep_lock_fs

 ... blah blah...
print DEBUG: overiding stale lock on $1 from pid $check

It used to work fine.. and I havent changed the code since zrep v0.8, it
looks like.
SO... something odd is going on on your system?

oh.

oh wait.
This may be a problem with the new zrep multi-overlaping process thing.
Because use of the new ZREPTAG thing.

As I said, previously, people complained that their system didnt have /proc

So, I worked around it, by storing the locking PID in a zfs property in the
filesystem.
property zrep:lock-pid

But now, potentialy multiple parallel zrep instances, could be squabbling
over the same global lock file. which at the moment, is truely global.

Hmmm....

**Comment by ppbrown** [comment]
For maximum paranoia and safety's sake, I am inclined to put in the limit
of,
"While you may have multiple zrep tags on a single filesystem.. you may
only run one instance of zrep at a time, on a particular single filesystem".

So if you had multiple destinations for some fs, you woud have to run, in
order

zrep -t dest1 -S data/store
# wait for finish, then
zrep -t dest2 -S data/store


how does that sit with you?

**Comment by darkpixel** [comment]
Ideally, I'd love to be able to schedule these in cron like:
```
0 2 * * * /usr/local/bin/zrep -t zrep-offsite sync all
0 * * * * /usr/local/bin/zrep -t zrep-local sync all
```

So sync to a local backup drive happens every hour on the hour, and off-site backups happen at some random time around midnight.

Ideally if the off-site backup ends up running for several hours, it wouldn't block the local backup.

I suppose some slightly creative crontabs would reduce the risk--say off-site backups start at 2 AM and on-site backups start at minute 0 for every hour except 2 AM and 3 AM.  But I'd hate to put that much thought into my cron jobs.  ;)  Especially since some remote sites have extremely slow connections where it could take ~6 hours to get the data off-site.

**Comment by ppbrown** [comment]
oh.

ick.

i forgot about disparate sync times and schedules.
A very valid need.

but unfortunately, one that is going to require some very careful coding.
I'm going to have to implement a multi-tier scheme of locking for that, I
think :-/
Phewww.......

**Comment by AndCycle** [comment]
ok, I hit similar issue, and found the cause,

my system is Gentoo Linux
I just fully upgrade my system and hit the problem,
it's cause by newer ksh version,

app-shells/ksh-93.20120229 is fine,
but app-shells/ksh-93.20140625 have this issue,

I upgrade my ksh due to old ver cannot compile by gcc 4.9,
I think at least you have to figure out what's the different in newer ksh caused this.

**Comment by ppbrown** [comment]
waitwhut.

thanks for the feedback, but... I'm thinking it was more a design flaw.
There's a race condition I thought.

**Comment by AndCycle** [comment]
I just look over the official ksh home page which doesn't touch anything since 2012,

so there is probably additional patch made by others,

https://dev.gentoo.org/~floppym/distfiles/
https://dev.gentoo.org/~floppym/distfiles/INIT.2014-06-25.tgz
https://dev.gentoo.org/~floppym/distfiles/ast-base.2014-06-25.tgz

if you interest about that, here is the path of those related source file,
I can simply use old binary and keep zrep happy for now.

**Comment by darkpixel** [comment]
I actually just finished doing a small bit of testing.  I don't believe it's a race condition.  I took zrep out of my crontab.  Then I ran a sync from the command line, made sure it worked, then enabled the local sync cron job only.  It bombs out with the same error.  I'm thinking the lock file gets left behind under one particular condition.  If I manually remove ```/var/run/zrep.lock``` and let cron run, it perform the sync one time only.  The second time it bombs out because the lock file exists.

I'll enable debugging next week when I'm in the office and see if I can pinpoint the problem.

**Comment by ppbrown** [comment]
But.. but.. it SHOULDNT work :-)

The same script, even when using different tags and a differnt proces, uses
the SAME lockfile.
And it uses lockfile-expired detection, based on tag, at present.
So I thought it should think, "oh its expired, I'll remov and continue",
when there is actually a currently running zrep using a differnt tag.

**Comment by darkpixel** [comment]
In my tests, there was only one instance of zrep running--the one started by cron.  The dataset I am syncing changes by a few megabytes an hour, so the sync runs in about 30 seconds (when I run it from the command line) and the job is scheduled to run every hour.  The first hour it runs, the second hour it complains about the lockfile.

Am I misunderstanding the process?

**Comment by ppbrown** [comment]
oh.

well thats a whole other issue maybe :-/
and THAT maybe due to bugs in your ksh

**Comment by darkpixel** [comment]
:beetle: Blast!

**Comment by AndCycle** [comment]
@darkpixel 
maybe you could paste info about your ksh ver?

my working ver is
ksh --version
sh (AT&T Research) 93u+ 2012-02-29

**Comment by darkpixel** [comment]
Both the sender and target are Debian:
```
root@usvansdnas01:~# ksh --version
  version         sh (AT&T Research) 93u+ 2012-08-01
root@usvansdnas01:~# 

root@uslog00nas03:~# ksh --version
  version         sh (AT&T Research) 93u+ 2012-08-01
root@uslog00nas03:~# 
```

**Comment by ppbrown** [comment]
my ksh is very old, since it's solaris ksh :)
it doesnt even have --version, for the solaris 10 ones!

solaris 11 uses
  version         sh (AT&T Research) 93u 2011-02-08

**Comment by AndCycle** [comment]
just downloaded 2012-08-01 and try it out on my system,
no, it won't cause problem on my side,

so there is something else happening here.

**Comment by darkpixel** [comment]
After more testing, I can repeatedly duplicate this through syncs kicked off from cron, but not from the command line.

crontab:
```
0 * * * * PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin HOME=/root /usr/local/bin/zrep -t zrep-local sync all
```

During testing I tried invoking it the same way (including setting the path) with no issues.  It's got to be something weird with the environment in cron...

**Comment by ppbrown** [comment]
Well, then you probably need to do

/bin/env >/tmp/cron.env

from cron, and compare result to when you run  env directly from command
line

**Comment by ppbrown** [comment]
OR, actually.. you might have clock skew problems, and cron jobs are
triggering twice.

but that really shouldnt happen....
your system would have to be seriously messed up to have its clock go
backwards.
Usually, systems just slow down the clock rather than back-stepping it, if
the difference to true time is small.


​

**Comment by darkpixel** [comment]
Setting it on the back burner for a few months made the problem magically disappear.  I was able to re-produce it on 24 machines when I first filed this.  Now I can't.  The only change was an upgrade to the latest zrep version.

**Comment by AndCycle** [comment]
@ppbrown 
I forget to mention, 

Gentoo is using ksh on https://github.com/att/ast master branch,
https://gitweb.gentoo.org/repo/gentoo.git/diff/app-shells/ksh/ksh-93.20160110.ebuild?id=396ed1a043ed4463735e018d7bf3b849e1f230ee

zrep v1.6.2 still have locking issue with the head version

**Comment by darkpixel** [comment]
Several of my machines have started showing this again, but I haven't had time to test/debug.

**Comment by ppbrown** [comment]
Please check latest release v1.6.7

**Comment by AndCycle** [comment]
tested over gentoo app-shells/ksh-93.20140625

```
# zrep
zrep v1.6.7: a program to replicate a zfs filesystem to another
```
```
# zrep sync all
ERROR: zrep_release_global_lock called, but do not own lock
Error: Cannot lock zmess/prod/iso. Cannot continue
```

still broken

**Comment by ppbrown** [comment]
Technically, that's a different error, though :)
Full details of the latest test would be nice.
along with running it in debug mode, with -v, or enabling DEBUG=1, i forget.

**Comment by AndCycle** [comment]
locking for single instance script execution is always a tough task
```
# zrep sync all
DEBUG: overiding stale lock on zmess/prod/iso from pid 2948
ERROR: zrep_release_global_lock called, but do not own lock
Error: Cannot lock zmess/prod/iso. Cannot continue
````

**Comment by AndCycle** [comment]
unrelated part, here is my `id`
```
# id
uid=0(root) gid=0(root) groups=0(root),1(bin),2(daemon),3(sys),4(adm),6(disk),10(wheel),11(floppy),26(tape),27(video),1005(hugetlb_shm_group)
```
you probably need to tweak the case match
```
case `id` in
    *\(root\))
    ZREP_RUNDIR=${ZREP_RUNDIR:-/var/run}
    ;;
    *)
    ZREP_RUNDIR=${ZREP_RUNDIR:-/tmp}
    ;;
esac

```

**Comment by AndCycle** [comment]
```
lockpid=`ls -l $Z_GLOBAL_LOCKFILE 2>/dev/null |awk -F/ '{print $NF}'`
```
my dig result the problem is on this line,
for some reason is spit out the directory content rather than just the `ln` pid path in the script

**Comment by AndCycle** [comment]
here is some dumb coding result, even I split the `ls` result to correct the error,
there are still problem due to parent pid issue
```
zrep_has_global_lock(){
    _debugprint zrep_has_global_lock 0 $Z_GLOBAL_LOCKFILE
    _debugprint zrep_has_global_lock 0.2 "ls -l $Z_GLOBAL_LOCKFILE 2>/dev/null |awk -F/ '{print $NF}'"
    ls_result=$(ls -l $Z_GLOBAL_LOCKFILE 2>/dev/null)
    lockpid=`echo $ls_result |awk -F/ '{print $NF}'`
#   lockpid=`ls -l $Z_GLOBAL_LOCKFILE 2>/dev/null |awk -F/ '{print $NF}'`

    _debugprint zrep_has_global_lock 1 $lockpid

    if [[ "$lockpid" == "" ]] ; then return 1 ; fi
    if [[ "$lockpid" != "$Z_GLOBAL_PID" ]] ; then
        _debugprint zrep_has_global_lock 2 not compare
        if [[ "$lockpid" != "$$" ]] ; then
            _debugprint zrep_has_global_lock 3 what $$
            return 1
        fi
    fi

    return 0
}
```
```
Failed to acquire global lock
Error: Cannot lock zmess/prod/iso. Cannot continue
DEBUG: zrep_has_global_lock 0 /tmp/zrep.lock
DEBUG: zrep_has_global_lock 0.2 ls -l /tmp/zrep.lock 2>/dev/null |awk -F/ '{print }'
DEBUG: zrep_has_global_lock 1 13754
DEBUG: zrep_has_global_lock 2 not compare
DEBUG: zrep_has_global_lock 3 what 14985
```

**Comment by AndCycle** [comment]
ok, maybe I shall reply once I really got to the bottom, gonna go to work
```
DEBUG: overiding stale lock on zmess/prod/iso from pid 13754
DEBUG: zrep_has_global_lock 0 /tmp/zrep.lock
DEBUG: zrep_has_global_lock 0.2 ls -l /tmp/zrep.lock 2>/dev/null |awk -F/ '{print }'
DEBUG: zrep_has_global_lock 1 -r--r--r-- 1 root root 0 Feb 24 15:19 schedstat -r--r--r-- 1 root root 0 Feb 24 15:19 sessionid -rw-r--r-- 1 root root 0 Feb 24 15:19 setgroups -r--r--r-- 1 root root 0 Feb 24 15:19 smaps -r--r--r-- 1 root root 0 Feb 24 15:19 stat -r--r--r-- 1 root root 0 Feb 24 15:19 statm -r--r--r-- 1 root root 0 Feb 24 15:19 status -r-------- 1 root root 0 Feb 24 15:19 syscall dr-xr-xr-x 3 root root 0 Feb 24 15:19 task -r--r--r-- 1 root root 0 Feb 24 15:19 timers -rw-rw-rw- 1 root root 0 Feb 24 15:19 timerslack_ns -rw-r--r-- 1 root root 0 Feb 24 15:19 uid_map -r--r--r-- 1 root root 0 Feb 24 15:19 wchan
DEBUG: zrep_has_global_lock 2 not compare
DEBUG: zrep_has_global_lock 3 what 22453
ERROR: zrep_release_global_lock called, but do not own lock
Error: Cannot lock zmess/prod/iso. Cannot continue
DEBUG: zrep_has_global_lock 0 /tmp/zrep.lock
DEBUG: zrep_has_global_lock 0.2 ls -l /tmp/zrep.lock 2>/dev/null |awk -F/ '{print }'
DEBUG: zrep_has_global_lock 1 -r--r--r-- 1 root root 0 Feb 24 15:19 schedstat -r--r--r-- 1 root root 0 Feb 24 15:19 sessionid -rw-r--r-- 1 root root 0 Feb 24 15:19 setgroups -r--r--r-- 1 root root 0 Feb 24 15:19 smaps -r--r--r-- 1 root root 0 Feb 24 15:19 stat -r--r--r-- 1 root root 0 Feb 24 15:19 statm -r--r--r-- 1 root root 0 Feb 24 15:19 status -r-------- 1 root root 0 Feb 24 15:19 syscall dr-xr-xr-x 3 root root 0 Feb 24 15:19 task -r--r--r-- 1 root root 0 Feb 24 15:19 timers -rw-rw-rw- 1 root root 0 Feb 24 15:19 timerslack_ns -rw-r--r-- 1 root root 0 Feb 24 15:19 uid_map -r--r--r-- 1 root root 0 Feb 24 15:19 wchan
DEBUG: zrep_has_global_lock 2 not compare
DEBUG: zrep_has_global_lock 3 what 22453
```

**Comment by darkpixel** [comment]
```locking for single instance script execution is always a tough task```

Follow the Unix philosophy.  Do one thing and do it well.  Let something like 'lockrun' take care of it.  (http://unixwiz.net/tools/lockrun.html)

**Comment by ppbrown** [comment]
I'd rather learn why things arent working on gentoo for you.
Some pieces of information are missing.

- when it says it cant get lock... is the lock actually held by a running
zrep process?
  - if so, what is the other proc  waiting on??
- your most recent output shows that zrep SUCCESSFULLY detected stale lock.
but wait...
 *  why would there be a stale lock?!?
 * how can it SUCCESSFULLY detect that it was stale there, but NOT in the
global lock code??

(Hmm.. did I USED to have cleanup for global,but it got lost?)
(Why do i have a comment there about problem with race condition??? sigh..)

**Comment by ppbrown** [comment]
Bah.

I figured out the race condition details I was worried about, from my code
comment in that section of the code.
(and added more comments ;)

So for now, i made an interim code change, that at least is a bit more
helpful when lock cant be grabbed due to stale:

+               if ! -d /proc/$lockpid ; then
+                       echo ERROR: stale global lock file >/dev/fd/2
+                       echo ERROR: shut down ALL zrep instances, then
manually
remove
+                       echo   $Z_GLOBAL_LOCKFILE
+               fi

Still worrying why there would be a stale global lockfile in the first
place!

**Comment by AndCycle** [comment]
that stale global lock is left over from last failed execution,
if I don't clean up /tmp/zrep.lock, it complains about that.

currently this issue all come after pateched newer version of ksh,
not the original ver. that haven't been update for years.

I wonder what ver of ksh do you use?

**Comment by ppbrown** [comment]
I use solaris standard:

$ ksh --version
  version         sh (AT&T Research) 93u 2011-02-08


If you can find out WHY it is failing, that would be very helpful.
(that is, the run that dies and leaves the global lockfile around, not the
one complaining that it cant get the lock)

**Comment by AndCycle** [comment]
that's a really safe old version, looks like the implement have changed a lot on beta branch,

after some rough exam, all issue occur around command substitution,
the return result is either gone or incorrect,
which is kind of mess...

**Comment by ppbrown** [comment]
aha.
well, not too surprised... it was a mess to try to get it actually working.
I had to blindly try combinations until I found something that worked.
Wouldnt surprise me if the working behaviour is something the maintainers
considered a bug, and "fixed" it.
Thanks So Much, meddlers :p

erm.. you're using a BETA version of ksh, did you say??

**Comment by ppbrown** [comment]
BTW, I recall that this is somewhat of a "Known Issue".

It's even documented by me:

http://www.bolthole.com/solaris/zrep/#compatibility

**Comment by AndCycle** [comment]
@ppbrown as I mentioned in earlier reply,
the version used on gentoo is https://github.com/att/ast/tree/beta

and forget all the nonsense I reply this few days,

after I pull your current master https://github.com/bolthole/zrep/commit/65c97beae1347c8be10c3000bad300338dccc3a9,
the major part with issue I currently found is
```
ls -l $Z_GLOBAL_LOCKFILE 2>/dev/null |awk -F/ '{print $NF}'
```
always return the directory content in ksh, need something protable,
it works if I replace it with `realpath` to resolve the symbolic link, but this is linux command,

and some error come out when there are `zrep.lock` left over
```
ln: /tmp/zrep.lock: cannot replace existing file
```
on linux `ln` won't overwrite without `-f`

**Comment by AndCycle** [comment]
do solaris have `pwd -P`?
`cd $Z_GLOBAL_LOCKFILE && pwd -P` 
can workaround this issue

**Comment by ppbrown** [comment]
Err.. what have you done??

$Z_GLOBAL_LOCKFILE...
is Supposed To Be A FILE.

Not a directory.

you have done very strange things.

and yes, the whole point of my using ln, is that I do not WANT it to
replace anything. That's what makes it safe to use :)

**Comment by ppbrown** [comment]
SIgh.
to clarify:

a symlink, is a file.
but in this case, the symlink is a symlink to a directory.

That shouldnt matter. the symlink itself is a FILE. "ls -l" should work on it...
UNLESS you have done something bad to ls, and somehow aliased it to "ls -F".

That is BAD. Dont Do That.

**Comment by ppbrown** [comment]
and on the side.. I dont see why the id checker doesnt match your "id" output.
it looks for "(root)", and it seems your id command does output "(root)". So I'm confused there.
ah. because I presumed it was not putting out group ids as well.
Added trailing '*' in latest version, should fix that.

**Comment by AndCycle** [comment]
this debug print doesn't make sense from the beginning.
and this is the only place need workaround, any better idea?

```
# /root/zrep-branch/zrep sync all
```
```
 #
 zrep_has_global_lock(){
        lockpid=`zrep_global_lock_pid`
+    _debugprint lockpid value $lockpid

        if [[ "$lockpid" == "" ]] ; then return 1 ; fi
        if [[ "$lockpid" != "$Z_GLOBAL_PID" ]] ; then
```
```
DEBUG: lockpid value total 0 -rw-r--r-- 1 root root 0 Feb 25 07:06 autogroup -r-------- 1 root root 0 Feb 25 07:06 auxv -r--r--r-- 1 root root 0 Feb 25 07:06 cgroup --w------- 1 root root 0 Feb 25 07:06 clear_refs -r--r--r-- 1 root root 0 Feb 25 07:06 cmdline -rw-r--r-- 1 root root 0 Feb 25 07:06 comm -rw-r--r-- 1 root root 0 Feb 25 07:06 coredump_filter -r--r--r-- 1 root root 0 Feb 25 07:06 cpuset root -r-------- 1 root root 0 Feb 25 07:06 environ ksh dr-x------ 2 root root 0 Feb 25 07:06 fd dr-x------ 2 root root 0 Feb 25 07:06 fdinfo -rw-r--r-- 1 root root 0 Feb 25 07:06 gid_map -r-------- 1 root root 0 Feb 25 07:06 io -r--r--r-- 1 root root 0 Feb 25 07:06 limits -rw-r--r-- 1 root root 0 Feb 25 07:06 loginuid dr-x------ 2 root root 0 Feb 25 07:06 map_files -r--r--r-- 1 root root 0 Feb 25 07:06 maps -rw------- 1 root root 0 Feb 25 07:06 mem -r--r--r-- 1 root root 0 Feb 25 07:06 mountinfo -r--r--r-- 1 root root 0 Feb 25 07:06 mounts -r-------- 1 root root 0 Feb 25 07:06 mountstats dr-xr-xr-x 8 root root 0 Feb 25 07:06 net dr-x--x--x 2 root root 0 Feb 25 07:06 ns -rw-r--r-- 1 root root 0 Feb 25 07:06 oom_adj -r--r--r-- 1 root root 0 Feb 25 07:06 oom_score -rw-r--r-- 1 root root 0 Feb 25 07:06 oom_score_adj -r-------- 1 root root 0 Feb 25 07:06 pagemap -r-------- 1 root root 0 Feb 25 07:06 personality -rw-r--r-- 1 root root 0 Feb 25 07:06 projid_map -r--r--r-- 1 root root 0 Feb 25 07:06 schedstat -r--r--r-- 1 root root 0 Feb 25 07:06 sessionid -rw-r--r-- 1 root root 0 Feb 25 07:06 setgroups -r--r--r-- 1 root root 0 Feb 25 07:06 smaps -r--r--r-- 1 root root 0 Feb 25 07:06 stat -r--r--r-- 1 root root 0 Feb 25 07:06 statm -r--r--r-- 1 root root 0 Feb 25 07:06 status -r-------- 1 root root 0 Feb 25 07:06 syscall dr-xr-xr-x 3 root root 0 Feb 25 07:06 task -r--r--r-- 1 root root 0 Feb 25 07:06 timers -rw-rw-rw- 1 root root 0 Feb 25 07:06 timerslack_ns -rw-r--r-- 1 root root 0 Feb 25 07:06 uid_map -r--r--r-- 1 root root 0 Feb 25 07:06 wchan
DEBUG: has_global_lock? no. lock held by PID total 0 -rw-r--r-- 1 root root 0 Feb 25 07:06 autogroup -r-------- 1 root root 0 Feb 25 07:06 auxv -r--r--r-- 1 root root 0 Feb 25 07:06 cgroup --w------- 1 root root 0 Feb 25 07:06 clear_refs -r--r--r-- 1 root root 0 Feb 25 07:06 cmdline -rw-r--r-- 1 root root 0 Feb 25 07:06 comm -rw-r--r-- 1 root root 0 Feb 25 07:06 coredump_filter -r--r--r-- 1 root root 0 Feb 25 07:06 cpuset root -r-------- 1 root root 0 Feb 25 07:06 environ ksh dr-x------ 2 root root 0 Feb 25 07:06 fd dr-x------ 2 root root 0 Feb 25 07:06 fdinfo -rw-r--r-- 1 root root 0 Feb 25 07:06 gid_map -r-------- 1 root root 0 Feb 25 07:06 io -r--r--r-- 1 root root 0 Feb 25 07:06 limits -rw-r--r-- 1 root root 0 Feb 25 07:06 loginuid dr-x------ 2 root root 0 Feb 25 07:06 map_files -r--r--r-- 1 root root 0 Feb 25 07:06 maps -rw------- 1 root root 0 Feb 25 07:06 mem -r--r--r-- 1 root root 0 Feb 25 07:06 mountinfo -r--r--r-- 1 root root 0 Feb 25 07:06 mounts -r-------- 1 root root 0 Feb 25 07:06 mountstats dr-xr-xr-x 8 root root 0 Feb 25 07:06 net dr-x--x--x 2 root root 0 Feb 25 07:06 ns -rw-r--r-- 1 root root 0 Feb 25 07:06 oom_adj -r--r--r-- 1 root root 0 Feb 25 07:06 oom_score -rw-r--r-- 1 root root 0 Feb 25 07:06 oom_score_adj -r-------- 1 root root 0 Feb 25 07:06 pagemap -r-------- 1 root root 0 Feb 25 07:06 personality -rw-r--r-- 1 root root 0 Feb 25 07:06 projid_map -r--r--r-- 1 root root 0 Feb 25 07:06 schedstat -r--r--r-- 1 root root 0 Feb 25 07:06 sessionid -rw-r--r-- 1 root root 0 Feb 25 07:06 setgroups -r--r--r-- 1 root root 0 Feb 25 07:06 smaps -r--r--r-- 1 root root 0 Feb 25 07:06 stat -r--r--r-- 1 root root 0 Feb 25 07:06 statm -r--r--r-- 1 root root 0 Feb 25 07:06 status -r-------- 1 root root 0 Feb 25 07:06 syscall dr-xr-xr-x 3 root root 0 Feb 25 07:06 task -r--r--r-- 1 root root 0 Feb 25 07:06 timers -rw-rw-rw- 1 root root 0 Feb 25 07:06 timerslack_ns -rw-r--r-- 1 root root 0 Feb 25 07:06 uid_map -r--r--r-- 1 root root 0 Feb 25 07:06 wchan
ERROR: zrep_release_global_lock called, but do not own lock
Error: Cannot lock zmess/prod/iso. Cannot continue
DEBUG: lockpid value total 0 -rw-r--r-- 1 root root 0 Feb 25 07:06 autogroup -r-------- 1 root root 0 Feb 25 07:06 auxv -r--r--r-- 1 root root 0 Feb 25 07:06 cgroup --w------- 1 root root 0 Feb 25 07:06 clear_refs -r--r--r-- 1 root root 0 Feb 25 07:06 cmdline -rw-r--r-- 1 root root 0 Feb 25 07:06 comm -rw-r--r-- 1 root root 0 Feb 25 07:06 coredump_filter -r--r--r-- 1 root root 0 Feb 25 07:06 cpuset root -r-------- 1 root root 0 Feb 25 07:06 environ ksh dr-x------ 2 root root 0 Feb 25 07:06 fd dr-x------ 2 root root 0 Feb 25 07:06 fdinfo -rw-r--r-- 1 root root 0 Feb 25 07:06 gid_map -r-------- 1 root root 0 Feb 25 07:06 io -r--r--r-- 1 root root 0 Feb 25 07:06 limits -rw-r--r-- 1 root root 0 Feb 25 07:06 loginuid dr-x------ 2 root root 0 Feb 25 07:06 map_files -r--r--r-- 1 root root 0 Feb 25 07:06 maps -rw------- 1 root root 0 Feb 25 07:06 mem -r--r--r-- 1 root root 0 Feb 25 07:06 mountinfo -r--r--r-- 1 root root 0 Feb 25 07:06 mounts -r-------- 1 root root 0 Feb 25 07:06 mountstats dr-xr-xr-x 8 root root 0 Feb 25 07:06 net dr-x--x--x 2 root root 0 Feb 25 07:06 ns -rw-r--r-- 1 root root 0 Feb 25 07:06 oom_adj -r--r--r-- 1 root root 0 Feb 25 07:06 oom_score -rw-r--r-- 1 root root 0 Feb 25 07:06 oom_score_adj -r-------- 1 root root 0 Feb 25 07:06 pagemap -r-------- 1 root root 0 Feb 25 07:06 personality -rw-r--r-- 1 root root 0 Feb 25 07:06 projid_map -r--r--r-- 1 root root 0 Feb 25 07:06 schedstat -r--r--r-- 1 root root 0 Feb 25 07:06 sessionid -rw-r--r-- 1 root root 0 Feb 25 07:06 setgroups -r--r--r-- 1 root root 0 Feb 25 07:06 smaps -r--r--r-- 1 root root 0 Feb 25 07:06 stat -r--r--r-- 1 root root 0 Feb 25 07:06 statm -r--r--r-- 1 root root 0 Feb 25 07:06 status -r-------- 1 root root 0 Feb 25 07:06 syscall dr-xr-xr-x 3 root root 0 Feb 25 07:06 task -r--r--r-- 1 root root 0 Feb 25 07:06 timers -rw-rw-rw- 1 root root 0 Feb 25 07:06 timerslack_ns -rw-r--r-- 1 root root 0 Feb 25 07:06 uid_map -r--r--r-- 1 root root 0 Feb 25 07:06 wchan
DEBUG: has_global_lock? no. lock held by PID total 0 -rw-r--r-- 1 root root 0 Feb 25 07:06 autogroup -r-------- 1 root root 0 Feb 25 07:06 auxv -r--r--r-- 1 root root 0 Feb 25 07:06 cgroup --w------- 1 root root 0 Feb 25 07:06 clear_refs -r--r--r-- 1 root root 0 Feb 25 07:06 cmdline -rw-r--r-- 1 root root 0 Feb 25 07:06 comm -rw-r--r-- 1 root root 0 Feb 25 07:06 coredump_filter -r--r--r-- 1 root root 0 Feb 25 07:06 cpuset root -r-------- 1 root root 0 Feb 25 07:06 environ ksh dr-x------ 2 root root 0 Feb 25 07:06 fd dr-x------ 2 root root 0 Feb 25 07:06 fdinfo -rw-r--r-- 1 root root 0 Feb 25 07:06 gid_map -r-------- 1 root root 0 Feb 25 07:06 io -r--r--r-- 1 root root 0 Feb 25 07:06 limits -rw-r--r-- 1 root root 0 Feb 25 07:06 loginuid dr-x------ 2 root root 0 Feb 25 07:06 map_files -r--r--r-- 1 root root 0 Feb 25 07:06 maps -rw------- 1 root root 0 Feb 25 07:06 mem -r--r--r-- 1 root root 0 Feb 25 07:06 mountinfo -r--r--r-- 1 root root 0 Feb 25 07:06 mounts -r-------- 1 root root 0 Feb 25 07:06 mountstats dr-xr-xr-x 8 root root 0 Feb 25 07:06 net dr-x--x--x 2 root root 0 Feb 25 07:06 ns -rw-r--r-- 1 root root 0 Feb 25 07:06 oom_adj -r--r--r-- 1 root root 0 Feb 25 07:06 oom_score -rw-r--r-- 1 root root 0 Feb 25 07:06 oom_score_adj -r-------- 1 root root 0 Feb 25 07:06 pagemap -r-------- 1 root root 0 Feb 25 07:06 personality -rw-r--r-- 1 root root 0 Feb 25 07:06 projid_map -r--r--r-- 1 root root 0 Feb 25 07:06 schedstat -r--r--r-- 1 root root 0 Feb 25 07:06 sessionid -rw-r--r-- 1 root root 0 Feb 25 07:06 setgroups -r--r--r-- 1 root root 0 Feb 25 07:06 smaps -r--r--r-- 1 root root 0 Feb 25 07:06 stat -r--r--r-- 1 root root 0 Feb 25 07:06 statm -r--r--r-- 1 root root 0 Feb 25 07:06 status -r-------- 1 root root 0 Feb 25 07:06 syscall dr-xr-xr-x 3 root root 0 Feb 25 07:06 task -r--r--r-- 1 root root 0 Feb 25 07:06 timers -rw-rw-rw- 1 root root 0 Feb 25 07:06 timerslack_ns -rw-r--r-- 1 root root 0 Feb 25 07:06 uid_map -r--r--r-- 1 root root 0 Feb 25 07:06 wchan
```

**Comment by AndCycle** [comment]
this is my suggestion, 
include this then current zrep master should be fine with `app-shells/ksh-93.20140625`


I can't figure out why `ls` got different behavior in this beta branch ksh,
the only alias for my `ls` is `alias ls='ls --color=auto'`

```
 # return PID of proc holding global lock, or nothing
 zrep_global_lock_pid(){
-       ls -l $Z_GLOBAL_LOCKFILE 2>/dev/null |awk -F/ '{print $NF}'
+    case `uname` in
+        Linux)
+        realpath $Z_GLOBAL_LOCKFILE  |awk -F/ '{print $NF}'
+        ;;
+        *)
+        ls -l $Z_GLOBAL_LOCKFILE 2>/dev/null |awk -F/ '{print $NF}'
+        ;;
+    esac
 }
```

**Comment by AndCycle** [comment]
```
bacztwo tmp # cat ./testksh.sh
#!/bin/ksh -p
alias
bacztwo tmp # ./testksh.sh
2d='set -f;_2d'
autoload='typeset -fu'
bool=_Bool
command='command '
compound='typeset -C'
fc=hist
float='typeset -lE'
functions='typeset -f'
hash='alias -t --'
history='hist -l'
integer='typeset -li'
nameref='typeset -n'
nohup='nohup '
r='hist -s'
redirect='command exec'
stop='kill -s STOP'
suspend='kill -s STOP $$'
times='{ { time;} 2>&1;}'
type='whence -v'
bacztwo tmp #
```

**Comment by ppbrown** [comment]
having "ls -l" not work on symlinks, IS A BUG (in gentoo ksh)

your patch isnt appropriate, since it already runs fine on every other
linux out there.
Just Not Gentoo.

here's what I suggest:
run

#!/bin/ksh -p

ln -s /var/tmp /tmp/symlink
echo proper ls
/bin/ls -l /tmp/symlink
echo built in ls
ls -l /tmp/symlink

#############

Then send the output to me, for reference. then also send the output to the
gentoo ksh ports maintainer, and point out,
"hey this is a bug. please fix this"

**Comment by AndCycle** [comment]
```
bacztwo tmp # ./ksh_ls
proper ls
lrwxrwxrwx 1 root root 8 Feb 25 15:17 /tmp/symlink -> /var/tmp
built in ls
total 35
drwxrwsr-x  19 root     portage        20 Feb 25 06:56 ccache
drwxr-xr-x   7 root     root           26 Feb 25 07:52 genkernel
drwxrwxr-x   4 portage  portage         4 Feb 25 07:21 portage
```

**Comment by AndCycle** [comment]
```
-P, --physical  Don't follow symbolic links. The default is determined by getconf PATH_RESOLVE.
```
```
#!/bin/ksh -p

ln -s /var/tmp /tmp/symlink
echo proper ls
/bin/ls -l /tmp/symlink
echo built in ls
ls -l /tmp/symlink
echo built in ls
ls -l -P /tmp/symlink
```
```
bacztwo tmp # ./ksh_ls
proper ls
lrwxrwxrwx 1 root root 8 Feb 25 15:20 /tmp/symlink -> /var/tmp
built in ls
total 36
drwxrwsr-x  19 root     portage        20 Feb 25 06:56 ccache
drwxr-xr-x   7 root     root           26 Feb 25 07:52 genkernel
drwxrwxr-x   4 portage  portage         4 Feb 25 07:21 portage
lrwxrwxrwx   1 root     root            8 Feb 25 15:19 tmp -> /var/tmp
built in ls
lrwxrwxrwx   1 root     root            8 Feb 25 15:20 /tmp/symlink -> /var/tmp
```

**Comment by ppbrown** [comment]
Ughh..
only gentoo is this difficult.
Unbutu doesnt do it.
Redhat doesnt do it.

But anyways... I read that it is possible to override this setting
with _AST_FEATURES
I would rather set a variable, than use "if"s.
So, if you play around and find the right value that works, I'll add it in.
Supposedly it is
export _AST_FEATURES="PATH_RESOLVE = physical"
but please let me know what really works.
I cant test it, because as I said: solaris, AND ubuntu AND redhat all use

version sh (AT&T Research) 93u+ 2012-08-01

**Comment by AndCycle** [comment]
just add the `--physical` flag to `ls` is sufficient, 

and gentoo is a system pull source code to compile as default,
from what I read in the build package on gentoo,
the only patch applied to current `ksh` on gentoo is to change some prefix map to the system,

by the way @darkpixel haven't talk about how current zrep master work on his system,
as he said he use debian,

maybe he can provide input on his side,

or you can pull current ksh beta ver and compile by yourself to verify the default.

**Comment by ppbrown** [comment]
that is not a valid solution, because the other OS's ls does not recognize
--physical
the ugliest alternative solution i would consider is to hardcode /bin/ls
instead of ls
but I'd rather use environment overrides through _AST_FEATURES, as I said.

**Comment by AndCycle** [comment]
you are the boss, I think you already fixed the issue that this post original encountered,
when deal with portability, it always ugly.

**Comment by ppbrown** [comment]
So, will you check for the correct values for _AST_FEATURES please?

**Comment by AndCycle** [comment]
you can just call `set` on your `ksh` I think,

here is mine.

```
#!/bin/ksh -p
set
```

```
_AST_FEATURES='CONFORMANCE - ast UNIVERSE - ucb PATH_RESOLVE - metaphysical'
```

and there something mentioned in changelog,
```
09-06-30 port/astconf.c: standard PATH_RESOLVE is "physical" (not "metaphysical")
```
long long times ago, the default is `physical` in `ksh`
then they changed
```
12-04-20 ls.c: FTW_META default
```
that make thing really complicated,

as some distribution prefer keep things work as it should like old day, they just patch it.

**Comment by AndCycle** [comment]
here is _AST_FEATURES I dump from one of my EC2 server
```
_AST_FEATURES='UNIVERSE - ucb'
```

and yea, it call system `ls` rather than built-in one as you concerned.

```
Linux ip-172-31-10-95 4.4.30-32.54.amzn1.x86_64 #1 SMP Thu Nov 10 15:52:05 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```
```
[ec2-user@ip-172-31-10-95 tmp]$ ./ksh_ls
proper ls
lrwxrwxrwx 1 ec2-user ec2-user 8 Feb 25 16:23 /tmp/symlink -> /var/tmp
built in ls
lrwxrwxrwx 1 ec2-user ec2-user 8 Feb 25 16:23 /tmp/symlink -> /var/tmp
built in ls
ls: invalid option -- 'P'
Try 'ls --help' for more information.
PATH_RESOLVE default
getconf: Unrecognized variable `PATH_RESOLVE'
```

**Comment by AndCycle** [comment]
http://unix.stackexchange.com/posts/200652/revisions
here is some trick you can you,
you can just test the `ls` is builtin or not.

**Comment by ppbrown** [comment]
I want to go for minimum change to environment.
doing "UNIVERSE - ucb" is a large change.

So, would you please test if just
_AST_FEATURES="PATH_RESOLVE - physical"
fixes it?

**Comment by AndCycle** [comment]
here is the trap, 
you have to set `_AST_FEATURES` before entering `ksh`,
which is too late for zrep,
as `PATH_RESOLVE` is a readonly variables report by `getconf`

**Comment by AndCycle** [comment]
```
# return PID of proc holding global lock, or nothing
zrep_global_lock_pid(){
    if `builtin ls`; then
        ls -l -P $Z_GLOBAL_LOCKFILE 2>/dev/null |awk -F/ '{print $NF}'
    else
        ls -l $Z_GLOBAL_LOCKFILE 2>/dev/null |awk -F/ '{print $NF}'
    fi
}
```

**Comment by ppbrown** [comment]
try:


if getconf PATH_RESOLVE 2>/dev/null ; then
  getconf PATH_RESOLVE = physical
fi
ls -l ...

**Comment by AndCycle** [comment]
after I tried your suggestion, 
it do inject getconf PATH_RESOLVE with physical,
but builtin `ls` doesn't care about it,

and that seems contaminate the entire system, now I have to find where do getconf store that value...

**Comment by ppbrown** [comment]
i was expecting you to test it as regular user, not root :-/

**Comment by ppbrown** [comment]
I just submitted a workaround build.
Effectively, makes it use /bin/ls just for gentoo.

**Comment by AndCycle** [comment]
if you just grab the source code and compile by yourself with
`sh bin/package flat only make ast-ksh SHELL=sh SHOPT_SYSRC=1`

you will see that `ksh` do carry a builtin`ls` with `PATH_RESOLVE=metaphysical`

it's much more complicate to me why everybody else won't get this,

this won't be a gentoo only issue, that's why I propose just detect is it a built-in ls.

**Comment by ppbrown** [comment]
SOMEday that may be a problem.
but right now, it's only a problem for gentoo.

hopefully even in the future, especially if you file a bug against gentoo,
other distros will override the new defaults, to make
PATH_RESOLVE=physical, to match the old standard behaviour.
it makes no sense to have a builtin ls, that has DIFFERENT functionality by
default, from the standard /bin/ls

**Comment by ppbrown** [comment]
I just read a claim that when you use getconf to change a value,

" Some getconf variables may be changed for the current process and its
descendants"

so.. that sounds like using it is supposed to be SAFE.

you previously seemed to claim that my test versions of zrep had
changed your system configuration.
Could you please double check that, and confirm or deny that
getconf PATH_RESOLVE physical

is actually safe to use for zrep?

**Comment by AndCycle** [comment]
my getconf is been meesed up with that, and reboot doesn't help at all,
now it even won't report correct value carried by ksh.

```
andcycle@bacztwo /tmp $ date
Thu Mar  2 12:06:08 CST 2017
andcycle@bacztwo /tmp $ cat test.sh
#!/bin/ksh -p

ls --version

ln -s /var/tmp /tmp/symlink
echo proper ls
/bin/ls -l /tmp/symlink
echo built in ls
ls -l /tmp/symlink
echo built in ls
ls -l -P /tmp/symlink
echo PATH_RESOLVE default
getconf PATH_RESOLVE
rm /tmp/symlink


andcycle@bacztwo /tmp $ ./test.sh
  version         ls (AT&T Research) 2013-12-06
proper ls
lrwxrwxrwx 1 andcycle andcycle 8 Mar  2 12:06 /tmp/symlink -> /var/tmp
built in ls
total 36
drwxrwsr-x  19 root     portage        20 Feb 28 16:55 ccache
drwxr-xr-x   7 root     root           26 Feb 25 15:38 genkernel
drwxrwxr-x   4 portage  portage         4 Mar  2 00:18 portage
lrwxrwxrwx   1 root     root            8 Feb 25 15:19 tmp -> /var/tmp
built in ls
lrwxrwxrwx   1 andcycle andcycle        8 Mar  2 12:06 /tmp/symlink -> /var/tmp
PATH_RESOLVE default
physical
andcycle@bacztwo /tmp $
```

**Comment by AndCycle** [comment]
and, by the way, the getconf used in ksh is also not the system one, it's builtin.

**Comment by ppbrown** [comment]
right, exactly.
very wierd.

maybe you should compare outputs from a non-ksh root shell. See if
PATH_RESOLVE got stuck in the GNU /usr/bin/getconf proggie. maybe that
could be doing something bad

**Comment by AndCycle** [comment]
no, PATH_RESOLVE only appear in ksh, is probably parsed from _AST_FEATURE,
the _AST_FEATURE is been messed as 
```
_AST_FEATURES='PATH_RESOLVE - physical CONFORMANCE - ast UNIVERSE - ucb'
```

my original _AST_FEATURE reported by ksh is
```
_AST_FEATURES='CONFORMANCE - ast UNIVERSE - ucb PATH_RESOLVE - metaphysical'
```

_AST_FEATURES is also in ksh only on my system,

no idea how to get those thing back to default.

**Comment by AndCycle** [comment]
and there are also some weird thing,
if I populated `getconf -w` first, the correct `PATH_RESOLVE` will be injected, but not permanent.

```
andcycle@bacztwo /tmp $ cat test3.sh
#!/bin/ksh -p
getconf PATH_RESOLVE
andcycle@bacztwo /tmp $ ./test3.sh
physical
```

```
andcycle@bacztwo /tmp $ cat test4.sh
#!/bin/ksh -p
getconf -w
getconf PATH_RESOLVE
andcycle@bacztwo /tmp $ ./test4.sh
ARCHITECTURE=x86_64
CONFORMANCE=ast
FS_3D=0
GETCONF=/usr/bin/getconf
HOSTTYPE=linux.i386
LIBPATH=lib:LD_LIBRARY_PATH
LIBPREFIX=lib
LIBSUFFIX=.so
PATH_ATTRIBUTES=0
PATH_RESOLVE=metaphysical
UNIVERSE=ucb
metaphysical

```

**Comment by AndCycle** [comment]
anyway it's too much effort for tackle this, there is lot's magic inside this part.

**Comment by ppbrown** [comment]
Notice how it says

GETCONF=/usr/bin/getconf


thats probably whats messing things up.

**Comment by ppbrown** [comment]
okay, so.. v1.6.8 has a hacky workaround for systems that have getconf PATH_RESOLVE.
(ie: gentoo :)
explicitly use /bin/ls instead of builtin.

A slight performance hit when checking for locks, and startup time (boo).
but teenytiny, so  I guess I can live with it.
Presuming this sucker is finally closed :)

**Comment by dansteen** [comment]
If it helps at all @ppbrown I was running into this issue on ArchLinux as well, and the 1.6.8 channel solved things as well.   (of course, then I forgot to sync with -R and seem to be messed up, but that's a different issue :-) )

Thanks for your work on this!

**Comment by ppbrown** [comment]
You gentlemen did some excellent detective work on all this.
That being said, I thought you might be interested to know that.... 
latest version of zrep no longer uses /bin/ls, so it is no longer needed ;-D
No more /proc usage. More portable.

Thanks for your interest in zrep!
